sessionspaces/src/main.rs

```diff
@@ -0,0 +1,162 @@
+#[derive(Debug, Default, derive_more::Deref)]
+struct SessionSpaces(BTreeMap<String, BTreeSet<String>>);
+
+impl SessionSpaces {
+    #[instrument(skip_all)]
+    fn new(sessions: Vec<Session>, subject_sessions: Vec<SubjectSession>) -> Self {
+        let mut spaces = BTreeMap::new();
+        for session in sessions.into_iter() {
+            spaces.insert(
+                session.id,
+                (
+                    format!("{}{}-{}", session.code, session.proposal, session.visit),
+                    BTreeSet::new(),
+                ),
+            );
+        }
+        for SubjectSession { subject, session } in subject_sessions.into_iter() {
+            if let Some(space) = spaces.get_mut(&session) {
+                space.1.insert(subject);
+            }
+        }
+        Self(spaces.into_values().collect())
+    }
+}
+
+/// Requests a new bundle from the bundle server and performs templating accordingly
+#[instrument(skip(k8s_client, current_sessions), err(level=tracing::Level::WARN))]
+async fn perform_update(
+    ispyb_pool: &MySqlPool,
+    k8s_client: kube::Client,
+    current_sessions: &mut SessionSpaces,
+) -> Result<(), anyhow::Error> {
+    info!("Fetching Sessions");
+    let sessions = Session::fetch(ispyb_pool).await?;
+    info!("Fetching Subjects");
+    let subjects = SubjectSession::fetch(ispyb_pool).await?;
+    let sessions = SessionSpaces::new(sessions, subjects);
+
+    let current_session_names = current_sessions.keys().cloned().collect::<BTreeSet<_>>();
+    let session_names = sessions.keys().cloned().collect::<BTreeSet<_>>();
+    let to_update = current_session_names
+        .union(&session_names)
+        .collect::<BTreeSet<_>>();
+
+    info!("Updating {} SessionSpaces", to_update.len());
+    for namespace in to_update.into_iter() {
+        let current_exists = current_sessions.contains_key(namespace);
+        let new_exists = sessions.contains_key(namespace);
+        let current_members = current_sessions.get(namespace);
+        let members = sessions.get(namespace);
+        if current_exists && !new_exists {
+            info!("Deleting Namespace: {}", namespace);
+            delete_namespace(namespace, k8s_client.clone()).await?;
+        } else if !current_exists && new_exists {
+            info!("Creating Namespace: {}", namespace);
+            create_namespace(namespace.clone(), k8s_client.clone()).await?;
+            create_argo_workflows_service_account(namespace.clone(), k8s_client.clone()).await?;
+            if let Some(members) = sessions.get(namespace) {
+                create_visit_member_service_account(
+                    namespace.clone(),
+                    members.clone(),
+                    k8s_client.clone(),
+                )
+                .await?;
+            }
+        } else if members.is_some() {
+            if current_members.is_none() || current_members.as_ref() != Some(&members.unwrap()) {
```

Comment:
Could use `is_some_and` here to avoid the `unwrap`

Reply:
makes sense!